internal/tracker/client.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2022 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package tracker
+
+import (
+	"context"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/tektoncd/pipeline/pkg/remote/oci"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type Client interface {
+	GetTektonObject(ctx context.Context, bundle, kind, name string) (runtime.Object, error)
+	GetImage(ctx context.Context, ref name.Reference) (v1.Image, error)
+}
+
+type contextKey string
+
+const clientContextKey contextKey = "ec.tracker.client"
+
+// WithClient returns a copy of given context with the included client.
+func WithClient(ctx context.Context, client Client) context.Context {
+	return context.WithValue(ctx, clientContextKey, client)
+}
+
+// NewClient returns the client from the context if set, otherwise a new instance.
+func NewClient(ctx context.Context) Client {
+	if client, ok := ctx.Value(clientContextKey).(Client); ok && client != nil {
+		return client
+	}
+	return exoClient{}
+}
+
+type exoClient struct{}
+
+func (c exoClient) GetTektonObject(ctx context.Context, bundle, kind, name string) (runtime.Object, error) {
```

Comment:
Can't this function and `GetImage` be called without exoClient? I'm not recognizing the need for the interface.

Reply:
It can. This is a pattern that we started using so the client implementation can be replaced with a different one. The use case in mind is supporting unit tests. This was inspired by [this article](https://incident.io/blog/golang-client-mocks). 

The alternative is to create global variables to functions that make an external call, and replace those during unit testing with something that returns mocked data. The problem with this approach, of course, is that it relies on global states which can have unintended side-effects.

Reply:
That makes sense. Thanks for the explanation. I've dealt with this by passing the functions to the struct that needs them, so I think in this case the `bundleInfo` struct would have fields `GetTektonObject` and `GetImage`, but I'm not sure how you'd test the `Track` method using that pattern.